internal/resource-request-operator/resourceRequest_controller.go

```diff
@@ -0,0 +1,162 @@
+	if resourceRequest.DeletionTimestamp.IsZero() {
+		// add finalizer
+		if !slice.ContainsString(resourceRequest.GetFinalizers(), finalizerName, nil) {
+			resourceRequest.SetFinalizers(append(resourceRequest.GetFinalizers(), finalizerName))
+			if err := r.Update(ctx, &resourceRequest); err != nil {
+				klog.Errorf("%s -> unable to update resource finalizer %s: %s", r.ClusterId, req.NamespacedName, err)
+				return ctrl.Result{}, err
+			}
+
+		}
+	} else {
+		//the resource is being deleted
+		if slice.ContainsString(resourceRequest.GetFinalizers(), finalizerName, nil) {
+
+			// remove our finalizer from the list and update it.
+			resourceRequest.SetFinalizers(slice.RemoveString(resourceRequest.GetFinalizers(), finalizerName, nil))
+			if err := r.Update(ctx, &resourceRequest); err != nil {
+				klog.Errorf("%s -> unable to update resource finalizer %s: %s", r.ClusterId, req.NamespacedName, err)
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	offer, offerErr := r.generateResourceOffer(&resourceRequest)
+	if offerErr != nil {
+		if offer != nil {
+			klog.Errorf("%s -> unable to generate resourceOffer %s", r.ClusterId, offer.Name)
+		}
+		klog.Errorf("%s -> Error generating resourceOffer: %s", r.ClusterId, offerErr)
+		return ctrl.Result{}, offerErr
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ResourceRequestReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&discoveryv1alpha1.ResourceRequest{}).
+		Complete(r)
+}
+
+//this function generate an empty offer
+func (r *ResourceRequestReconciler) generateResourceOffer(request *discoveryv1alpha1.ResourceRequest) (*sharingv1alpha1.ResourceOffer, error) {
+	offerUID := uuid.NewUUID()
+	err := r.computeResources()
+	if err != nil {
+		return nil, err
+	}
+	offer := &sharingv1alpha1.ResourceOffer{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: request.GetNamespace(),
+			Name:      offerPrefix + r.ClusterId,
+			UID:       offerUID,
+			Labels: map[string]string{
+				discovery.ClusterIdLabel: request.Spec.ClusterIdentity.ClusterID,
+			},
+		},
+		Spec: sharingv1alpha1.ResourceOfferSpec{
+			ClusterId: r.ClusterId,
+			Images:    []corev1.ContainerImage{},
+			LimitRange: corev1.LimitRangeSpec{
+				Limits: []corev1.LimitRangeItem{},
+			},
+			ResourceQuota: corev1.ResourceQuotaSpec{
+				Hard: resources.Offers,
+			},
+			KubeConfigRef: corev1.SecretReference{
+				Namespace: "liqo",
+				Name:      "name",
+			},
+			Timestamp:  metav1.NewTime(time.Now()),
+			TimeToLive: metav1.NewTime(time.Now().Add(30 * time.Minute)),
+		},
+	}
+
+	//set owner reference
+	err = controllerutil.SetOwnerReference(request, offer, r.Scheme)
+	if err != nil {
+		klog.Errorf("%s -> unable to set owner reference on resourceRequest %s with offer %s", r.ClusterId, request.Name, offer.Name)
+		return nil, err
+	}
+
+	op, err := controllerutil.CreateOrUpdate(context.Background(), r.Client, offer, func() error {
+		return nil
+	})
```

Comment:
This is not the correct usage of the CreateOrUpdate function, as it works only for the creation. See [the documentation](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.0/pkg/controller/controllerutil#CreateOrUpdate) for more information